File: libgomp/testsuite/libgomp.fortran/udr3.f90
! { dg-do run }

!$omp declare reduction (foo : character(kind=1, len=*) &
!$omp & : omp_out = trim(omp_out) // omp_in) initializer (omp_priv = '')
!$omp declare reduction (bar : character(kind=1, len=:) &
!$omp & : omp_out = trim(omp_in) // omp_out) initializer (omp_priv = '')
!$omp declare reduction (baz : character(kind=1, len=1) &
!$omp & : omp_out = char (ichar (omp_out) + ichar (omp_in) &
!$omp & - ichar ('0'))) initializer (omp_priv = '0')
!$omp declare reduction (baz : character(kind=1, len=2) &
!$omp & : omp_out = char (ichar (omp_out(1:1)) + ichar (omp_in(1:1)) &
!$omp & - ichar ('0')) // char (ichar (omp_out(2:2)) + &
!$omp & ichar (omp_in(2:2)) - ichar ('0'))) initializer (omp_priv = '00')
  character(kind=1, len=64) :: c, d
  character(kind = 1, len=1) :: e
  character(kind = 1, len=1+1) :: f
  integer :: i
  c = ''
  d = ''
  e = '0'
  f = '00'
!$omp parallel do reduction (foo : c) reduction (bar : d) &
!$omp & reduction (baz : e, f)
  do i = 1, 64
    c = trim(c) // char (ichar ('0') + i)
    d = char (ichar ('0') + i) // d
    e = char (ichar (e) + mod (i, 3))
    f = char (ichar (f(1:1)) + mod (i, 2)) &
&	// char (ichar (f(2:2)) + mod (i, 3))
  end do
  do i = 1, 64
    if (index (c, char (ichar ('0') + i)) .eq. 0) call abort
    if (index (d, char (ichar ('0') + i)) .eq. 0) call abort
  end do
  if (e.ne.char (ichar ('0') + 64)) call abort
  if (f(1:1).ne.char (ichar ('0') + 32)) call abort
  if (f(2:2).ne.char (ichar ('0') + 64)) call abort
end

File: gcc/testsuite/gcc.dg/gomp/udr-3.c
/* { dg-do compile } */
/* { dg-options "-fopenmp -std=gnu89" } */

struct S { int s; };
struct T { int t; };
struct U { int u; };

#pragma omp declare reduction (+: struct S: omp_out.s += omp_in.s)
#pragma omp declare reduction (*: struct S: omp_out.s *= omp_in.s) \
		    initializer (omp_priv = {1})
#pragma omp declare reduction (foo: struct S: omp_out.s += omp_in.s)

void
f1 ()
{
  struct S s, s2;
  struct T t;
  #pragma omp declare reduction (+: struct T: omp_out.t += omp_in.t)
  #pragma omp parallel reduction (+: t) reduction (foo: s) reduction (*: s2)
  s.s = 1, t.t = 1, s2.s = 2;
  #pragma omp parallel reduction (+: s)
  s.s = 1;
}

void bar (struct S *);

void
f2 ()
{
  #pragma omp declare reduction (foo: struct S: omp_out.s += omp_in.s) initializer (bar (&omp_priv))
  #pragma omp declare reduction (bar: struct S: omp_out.s += omp_in.s) initializer (bar (&omp_orig)) /* { dg-error "one of the initializer call arguments should be" } */
}

#pragma omp declare reduction (+: struct U: omp_out.u *= omp_in.u)		/* { dg-error "previous" } */
#pragma omp declare reduction (+: struct U: omp_out.u += omp_in.u)		/* { dg-error "redeclaration of" } */

void
f3 ()
{
  #pragma omp declare reduction (f3: struct U: omp_out.u *= omp_in.u)		/* { dg-error "previous" } */
  #pragma omp declare reduction (f3: struct U: omp_out.u += omp_in.u)		/* { dg-error "redeclaration of" } */
}

struct V
{
  #pragma omp declare reduction (bar: struct S: omp_out.s *= omp_in.s)		/* { dg-error "not at file or block scope" } */
  #pragma omp declare reduction (bar: struct S: omp_out.s += omp_in.s)		/* { dg-error "not at file or block scope" } */
};

#pragma omp declare reduction (n3: long: omp_out += omp_in)		/* { dg-error "previous" } */
#pragma omp declare reduction (n3: long int: omp_out += omp_in)		/* { dg-error "redeclaration of" } */
#pragma omp declare reduction (n3: short unsigned: omp_out += omp_in)
#pragma omp declare reduction (n3: short int: omp_out += omp_in)

void
f4 (void)
{
  #pragma omp declare reduction (f4: long: omp_out += omp_in)		/* { dg-error "previous" } */
  #pragma omp declare reduction (f4: long int: omp_out += omp_in)	/* { dg-error "redeclaration of" } */
  #pragma omp declare reduction (f4: short unsigned: omp_out += omp_in)
  #pragma omp declare reduction (f4: short int: omp_out += omp_in)
}

void
f5 (void)
{
  #pragma omp declare reduction (+: struct S: omp_out.s += omp_in.s) initializer (omp_priv) /* { dg-error "expected" } */
  #pragma omp declare reduction (+: struct T: omp_out.t += omp_in.t) initializer (omp_priv ()) /* { dg-error "expected" } */
}

void
f6 (a, b)
#pragma omp declare reduction (bar: struct S: omp_out.s *= omp_in.s)	/* { dg-error "expected declaration specifiers before" } */
  int a;
  int b;
{
}
